lib/TelemetryTracker.js

```diff
@@ -0,0 +1,120 @@
+  "RECENT_BOOKMARKS_RESPONSE",
+  "RECENT_LINKS_REQUEST",
+  "RECENT_LINKS_RESPONSE",
+  "FRECENT_LINKS_REQUEST",
+  "FRECENT_LINKS_RESPONSE",
+  "SEARCH_STATE_REQUEST",
+  "SEARCH_STATE_RESPONSE",
+  "NOTIFY_TELEMETRY",
+]);
+
+function TelemetryTracker(trackableURLs) {
+  this._trackableURLs = trackableURLs;
+  this.init();
+}
+
+TelemetryTracker.prototype = {
+
+  get events() {
+    return this._events;
+  },
+
+  init: function() {
+    this.onOpen = this.onOpen.bind(this);
+    this.onReady = this.onReady.bind(this);
+    this.onPrefChange = this.onPrefChange.bind(this);
+    if (simplePrefs.prefs.telemetryLog) {
+      tabs.on("open", this.onOpen);
+      this._active = true;
+    }
+    simplePrefs.on("telemetryLog", this.onPrefChange);
+  },
+
+  uninit: function() {
+    if (simplePrefs.prefs.telemetryLog) {
+      tabs.removeListener("open", this.onOpen);
+    }
+    simplePrefs.removeListener("telemetryLog", this.onPrefChange);
+  },
+
+  onPrefChange: function() {
+    if (simplePrefs.prefs.telemetryLog) {
+      tabs.on("open", this.onOpen);
+      this._active = true;
+    } else {
+      tabs.removeListener("open", this.onOpen);
+      this._active = false;
+      this.clear();
+    }
+    Services.obs.notifyObservers(null, "telemetry-pref-changed", null);
+  },
+
+  isActivityStreamsURL: function(URL) {
+    return this._trackableURLs.indexOf(URL) !== -1;
+  },
+
+  onReady: function(tab) {
+    if (this.isActivityStreamsURL(tab.url)) {
+      // starting new telemetry events collection
+      this.clear();
+      this.log(tab, "TAB_READY");
+    }
+    tab.removeListener("ready", this.onReady);
+  },
+
+  onOpen: function(tab) {
+    // since we don't know the url, timestamp the event
+    this._lastTabOpen = Date.now();
+    tab.on("ready", this.onReady);
+  },
+
+  log: function(tab, tag, data) {
+    if (this._active && this.isActivityStreamsURL(tab.url) && VALID_TELEMETRY_TAGS.has(tag)) {
+      if (!this._events) {
+        this._events = [];
+        this._events.push({tag: "TAB_OPEN", start: this._lastTabOpen});
+      }
+      this._events.push({tag: tag, start: Date.now(), data: data});
+      if (data == "NEWTAB_RENDER") {
+        this.report();
+        Services.obs.notifyObservers(null, "telemetry-complete", JSON.stringify(this.events));
+      }
+    }
+  },
+
+  report: function() {
+    let tagTimes = new Map();
+    let firstStart = this._events[0].start;
+    this._events.forEach(item => {
+      // subtract first start time from all subsequent times
+      item.start -= firstStart;
+      if (item.tag.endsWith("_RESPONSE")) {
+        // if response, record the time delta
+        let request = item.tag.replace("_RESPONSE", "_REQUEST");
+        item.delta = item.start - tagTimes.get(request);
+      }
+      // record start time for the event
+      tagTimes.set(item.tag, item.start);
+      console.log(item.start + " " + item.tag + (item.data ? ":" + item.data : "") + " " + (item.delta || ""));
```

Comment:
This should print to the terminal output (STDOUT if running jpm, otherwise devtools).

The eslint rule for console statements should be removed; debug logs are a useful tool.